Add `.specific()` page queryset method

timheap opened the pull request:
This method gets the specific instances of all pages in a queryset in an efficient manner, hitting the database the minimum number of times.

timheap commented:
The drone build seems to be failing for an unrelated error...

gasman commented:
The Drone error is from #1244. Will put in a fix/workaround for that now - even if it's a server configuration issue, it could come up elsewhere, and we don't want to put in unnecessary roadblocks to people running the tests...

kaedroho commented:
Thanks @timheap!

It looks like ``.specific()`` will evaluate the entire queryset which might not be great for performance if you want to paginate a list of thousands of specific pages, for example.

Ideally, ``.specific()`` would return a lazily-evaluated list of pages (much like ``ValuesQuerySet`` and ``ElasticsSearchResults``). This allows the user to slice the QuerySet before it is evaluated so only the pages that are wanted are retrieved.

timheap commented:
You are correct that it evaluates the whole queryset when called. It is possible to filter and slice the queryset before calling `.specific()` though, which allows you to do sensible limiting of the results before calling it.

I did not think of doing something like a `ValuesQuerySet` though. I had a quick look in to that, and it seems quite easy to do. I will investigate further.

timheap commented:
Actually, looking in to it further, the way `.values()` works in Django is about to change *significantly*. 

Version 1.8 and below have a very verbose method that changes the base QuerySet class to a generated subclass of the model manager and a ValuesQuerySet or similar.

The current master of Django has a much cleaner way of doing it ([Django #2421](https://code.djangoproject.com/ticket/24211)) that just sets an iterator class for the QuerySet to a `ValuesIterator` or similar. Writing a new iterator appears simple.

Supporting both versions should be possible, although it will likely be a little messy

timheap commented:
I made it work in a similar manner to a ValuesQuerySet. The `.specific()` call is now lazy, and you can continue to use the QuerySet after a call to `.specific()`.

I added Django 1.9 compatible code, but Wagtail does not yet run on Django 1.9 (something in django-taggit breaks) so I have no idea if it works.

kaedroho commented:
Just tested, seems to work really well. Good work!

One cool about implementing it as a ``QuerySet`` is that it works nicely with search (when using the elasticsearch backend):

``
>>> Page.objects.all().specific().search("Ev")
[<EventPage: Event 1>, <EventPage: Event 2>, <EventIndexPage: Events index>]
``

kaedroho commented:
I'm happy for this to be merged as soon as the 1.1 merge window opens. @gasman would probably want to look at it though.

gasman commented on wagtail/wagtailcore/query.py:
```diff
@@ -206,3 +209,60 @@ def unpublish(self):
+        using the minimum number of queries.
+
+        As this returns a list, no further queries can be chained after a call
+        to `specific`.
```
I guess this sentence is no longer true...?

gasman commented:
This looks great - I'm sure Wagtail devs everywhere will welcome this!

Only nitpick I have is that the top-level `specific` function https://github.com/torchbox/wagtail/pull/1251/files#diff-97474ebd551e3ac9c34c55efa4a00b19R230 ought to have a docstring to explain its purpose, and possibly be given a more descriptive name.

timheap commented:
I've updated/added docstrings, and renamed the function to `specific_iterator`. I've also fleshed out the documentation with a link to `Page.specific`.

gasman commented:
Thanks! Now committed in 7d7eece0d1bc1f9b04fc7c28e260e2ec00f4e15a (looks like you pushed the updated version to takeflight/master rather than this branch, so I've cherry-picked it from there...)

timheap commented:
Whoops, you're right, I pushed to the wrong location. Thanks for catching that, and thanks for merging.
